plugins/catalog/src/components/CatalogPage/CatalogTabs.tsx

```diff
@@ -14,42 +14,72 @@
  * limitations under the License.
  */
 
-import React, { useCallback, useEffect } from 'react';
-import { HeaderTabs } from '@backstage/core';
-import { labeledEntityTypes, LabeledEntityType } from '../../data/filters';
-import { useEntityFilterGroup, FilterGroup } from '../../filter';
 import { Entity } from '@backstage/catalog-model';
+import { HeaderTabs } from '@backstage/core';
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import { FilterGroup, useEntityFilterGroup } from '../../filter';
 
-const filterGroup: FilterGroup = {
-  filters: Object.fromEntries(
-    labeledEntityTypes.map(t => [
-      t.id,
-      (entity: Entity) => entity.spec?.type === t.id,
-    ]),
-  ),
+/**
+ * A component type, and a human readable label for it.
+ */
+export type LabeledComponentType = {
+  id: string;
+  label: string;
 };
 
+/**
+ * Called on mount, and when the selected tab changes.
+ */
+export type OnChangeCallback = (tab: LabeledComponentType) => void;
+
 type Props = {
-  onChange?: (type: LabeledEntityType) => void;
+  tabs: LabeledComponentType[];
+  onChange?: OnChangeCallback;
 };
 
-export const CatalogTabs = ({ onChange }: Props) => {
+/**
+ * The tabs at the top of the catalog list page, for component type filtering.
+ */
+export const CatalogTabs = ({ tabs, onChange }: Props) => {
+  const filterGroup = useMemo<FilterGroup>(() => {
+    return {
+      filters: Object.fromEntries(
+        tabs.map(t => [t.id, (entity: Entity) => entity.spec?.type === t.id]),
+      ),
+    };
+  }, [tabs]);
+
   const { setSelectedFilters } = useEntityFilterGroup('type', filterGroup, [
-    labeledEntityTypes[0].id,
+    tabs[0].id,
   ]);
 
-  const onChangeFn = useCallback(
-    (index: Number) => {
-      const type = labeledEntityTypes[index as number];
-      setSelectedFilters([type.id]);
-      onChange?.(type);
-    },
-    [onChange, setSelectedFilters],
-  );
+  const [currentTabIndex, setCurrentTabIndex] = useState<number>(0);
 
+  // Hold a reference to the callback
+  const onChangeRef = useRef<OnChangeCallback>();
   useEffect(() => {
-    onChange?.(labeledEntityTypes[0]);
+    onChangeRef.current = onChange;
   }, [onChange]);
 
-  return <HeaderTabs tabs={labeledEntityTypes} onChange={onChangeFn} />;
+  useEffect(() => {
+    onChangeRef.current?.(tabs[currentTabIndex]);
+  }, [tabs, currentTabIndex]);
+
+  const switchTab = useCallback(
+    (index: number) => {
+      const tab = tabs[index];
+      setSelectedFilters([tab.id]);
+      setCurrentTabIndex(index);
+      onChangeRef.current?.(tab);
+    },
+    [tabs, setSelectedFilters],
```

Comment:
think you should probably add `onChangeRef?.current` and mauybe `setSetSelectedFilters` to the `deps` array

Reply:
we should have linting rules around this.

Reply:
We have exhaustive-deps but actually refs are explicitly exempt, and that's the reason I use them.

The reason for the whole ref dance for the callbacks is that I want the caller to be able to render `<MyComponent onCallback={() => { somecode()}} />` which generates an entirely new function object on each render. If we use that in a dep array for something that does setState we will end up in infinite render loops. And the callbacks really aren't part of any real state if you think of it, conceptually, so a ref makes sense

Reply:
Oh and I think you meant `setCurrentTabIndex` for the other one? `useState` setters (which are static) are exempt too.

Reply:
ah ok. makes sense